xline/src/server/watch_server.rs

```diff
@@ -301,31 +289,46 @@ where
 
     /// Handle watch event
     async fn handle_watch_event(&mut self, mut watch_event: WatchEvent) {
-        let mut events = watch_event.take_events();
-        if events.is_empty() {
-            return;
-        }
         let watch_id = watch_event.watch_id();
-        if self.prev_kv.contains(&watch_id) {
-            for ev in &mut events {
-                if !ev.is_create() {
-                    let kv = ev
-                        .kv
-                        .as_ref()
-                        .unwrap_or_else(|| panic!("event.kv can't be None"));
-                    ev.prev_kv = self.kv_watcher.get_prev_kv(kv);
+        let response = if watch_event.compacted() {
+            WatchResponse {
+                header: Some(ResponseHeader {
+                    revision: watch_event.revision(),
+                    ..ResponseHeader::default()
+                }),
+                watch_id,
+                compact_revision: self.kv_watcher.compacted_revision(),
+                canceled: true,
+                ..WatchResponse::default()
+            }
+        } else {
+            let mut events = watch_event.take_events();
+            if events.is_empty() {
+                return;
+            }
+
+            if self.prev_kv.contains(&watch_id) {
+                for ev in &mut events {
+                    if !ev.is_create() {
+                        let kv = ev
+                            .kv
+                            .as_ref()
+                            .unwrap_or_else(|| panic!("event.kv can't be None"));
+                        ev.prev_kv = self.kv_watcher.get_prev_kv(kv);
+                    }
                 }
             }
-        }
-        let response = WatchResponse {
-            header: Some(ResponseHeader {
-                revision: watch_event.revision(),
-                ..ResponseHeader::default()
-            }),
-            watch_id,
-            events,
-            ..WatchResponse::default()
+            WatchResponse {
+                header: Some(ResponseHeader {
+                    revision: watch_event.revision(),
+                    ..ResponseHeader::default()
+                }),
+                watch_id,
+                events,
+                ..WatchResponse::default()
+            }
         };
```

Comment:
Please build `WatchResponse` before `if else` block

Reply:
How come?  The `if else` block is responsible for returning different `WatchResponse` based on whether the given `watch_event` is compacted or not.

Reply:
we can build the response first, and modify it in different branches to reduce code duplication

Reply:
Done